sasl/client.go

```diff
@@ -0,0 +1,298 @@
+package sasl
+
+import (
+	"crypto/md5"
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+)
+
+const (
+	saslMD5NoneSize = 32
+	saslDataMax     = 65536
+
+	sealingClientServer = "Digest H(A1) to client-to-server sealing key magic constant"
+	sealingServerClient = "Digest H(A1) to server-to-client sealing key magic constant"
+	signingClientServer = "Digest session key to client-to-server signing key magic constant"
+	signingServerClient = "Digest session key to server-to-client signing key magic constant"
+)
+
+type saslInfoCallback func([]CallbackRequestItem) error
+
+type context struct {
+	State int
+	challenge
+	KIReceive []byte
+	KISend    []byte
+	EncKey    []byte
+	DecKey    []byte
+	HA1       []byte
+}
+
+type client struct {
+	LocalAddress  net.Addr
+	RemoteAddress net.Addr
+	Callback      saslInfoCallback
+	context       *context
+}
+
+type challenge struct {
+	Service    string
+	Hostname   string
+	DigestURI  string
+	Username   string
+	AuthzID    string
+	NonceHex   string
+	CNonceHex  string
+	NC         uint32
+	Algorithm  string
+	QOP        string
+	Realm      string
+	Charset    string
+	MaxBuf     uint32
+	Response   string
+	Method     string
+	SessionKey string
+	Ciphers    []string
+	Cipher     string
+	CipherType cipherType
+}
+
+// Client represents the client portion of the SASL protocol
+type Client interface {
+	FindMech(mechList []string) (mech string, err error)
+	ApplyChallenge(challenge string)
+	MakeResponse() string
+	SetCNonce(string)
+}
+
+// NewClient creates a new instance of the client
+func NewClient(service, hostname string, auth AuthInfo) Client {
+	nonce := make([]byte, saslMD5NoneSize)
+	rand.Read(nonce)
+	return &client{
+		Callback: auth.Callback,
+		context: &context{
+			State: 1,
+			challenge: challenge{
+				Service:   service,
+				Hostname:  hostname,
+				Realm:     "",
+				NC:        0,
+				CNonceHex: base64.StdEncoding.EncodeToString(nonce),
+				MaxBuf:    saslDataMax,
+				Charset:   "utf-8",
+				Method:    "AUTHENTICATE",
+			},
+		},
+	}
+}
+
+func (client *client) FindMech(mechList []string) (mech string, err error) {
+	for _, entry := range mechList {
+		// Eventually this would be the place to switch to different
+		// mech variations, but for now we only support md5
+		if entry == "DIGEST-MD5" {
+			return entry, nil
+		}
+	}
+	return "", fmt.Errorf("SASL: Requested mech not supported")
```

Comment:
Lowercase `sasl` at start of error message.  Maybe toss in a `%v` to show the invalid `mechList`.